metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractValueCodeAction.scala

```diff
@@ -33,55 +33,131 @@ class ExtractValueCodeAction(
     val path = params.getTextDocument().getUri().toAbsolutePath
     val range = params.getRange()
 
-    val applyOpt = trees.findLastEnclosingAt[Term.Apply](
-      path,
-      range.getStart(),
-      appl => appl.args.exists(_.pos.encloses(range))
-    )
-
-    val textEdits = for {
-      apply <- applyOpt
-      stats <- lastEnclosingStatsList(apply)
-      argumentOpt <- apply.args.find { arg => arg.pos.encloses(range) }
-      argument = argumentOpt match {
-        // named paramaeter
-        case Term.Assign(_, rhs) => rhs
-        case other => other
+    val allTrees: Seq[Term] = {
+      def loop(t: Tree): Seq[Term] = {
+        t.children.find(_.pos.encloses(range)) match {
+          case Some(tr: Term) if existsRangeEnclosing(tr, range) =>
+            loop(tr) :+ tr
+          case Some(tr) => loop(tr)
+          case None => Nil
+        }
+      }
+      trees.get(path).map(loop(_)).getOrElse(Nil)
+    }
+    val textEdits: Option[Seq[(Seq[l.TextEdit], String)]] = {
+      for {
+        head <- allTrees.headOption
+        stats <- lastEnclosingStatsList(head)
+        name = createNewName(stats)
+        stat <- stats.find(stat => stat.pos.encloses(head.pos))
+        source <- buffers.get(path)
+      } yield {
+        val blank =
+          if (source(stat.pos.start - stat.pos.startColumn) == '\t') '\t'
+          else ' '
+        val indent = blank.stringRepeat(indentationLength(source, stat.pos))
+        val keyword = if (stat.isInstanceOf[Enumerator]) "" else "val "
+        // we will insert `val newValue = ???` before the existing statement containing apply
+        for {
+          term <- allTrees
+          argument <- findRangeEnclosing(term, range)
+          // avoid extracting lambdas (this needs actual type information)
+          if isNotLambda(argument)
+        } yield {
+          val valueText = s"${indent}$keyword$name = ${argument.toString()}"
+          val valueTextWithBraces = withBraces(stat, source, valueText, blank)
+          // we need to add additional () in case of  `apply{}`
+          val replacementText =
+            term match {
+              case apply: Term.Apply
+                  if argument.is[Term.Block] && !applyHasParens(apply) =>
+                s"($name)"
+              case _ => name
+            }
+          val replacedArgument =
+            new l.TextEdit(argument.pos.toLSP, replacementText)
+          (valueTextWithBraces :+ replacedArgument, argument.toString())
```

Comment:
Similarly here, `:+` -> `::`. 
Speaking about List and Seq, I think `withBraces` can have explicit `List` type and `Seq` constructors can be replaced by `List` one.